processor/spanmetricsprocessor/processor.go

```diff
@@ -141,15 +193,187 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		dpLatency.SetStartTime(pdata.TimeToUnixNano(p.startTime))
+		dpLatency.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
+		dpLatency.SetExplicitBounds(p.latencyBounds)
+		dpLatency.SetBucketCounts(p.latencyBucketCounts[key])
+		dpLatency.SetCount(p.latencyCount[key])
+		dpLatency.SetSum(int64(p.latencySum[key]))
+
+		dpLatency.LabelsMap().InitFromMap(p.metricKeyToDimensions[key])
+
+		mLatency := pdata.NewMetric()
+		mLatency.SetDataType(pdata.MetricDataTypeIntHistogram)
+		mLatency.SetName("latency")
+		mLatency.IntHistogram().DataPoints().Append(dpLatency)
+		mLatency.IntHistogram().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mLatency)
+	}
+}
+
+// collectCallMetrics collects the raw call count metrics, writing the data
+// into the given instrumentation library metrics.
+func (p *processorImp) collectCallMetrics(ilm *pdata.InstrumentationLibraryMetrics) {
+	for key := range p.callSum {
+		dpCalls := pdata.NewIntDataPoint()
+		dpCalls.SetStartTime(pdata.TimeToUnixNano(p.startTime))
+		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
+		dpCalls.SetValue(p.callSum[key])
+
+		dpCalls.LabelsMap().InitFromMap(p.metricKeyToDimensions[key])
+
+		mCalls := pdata.NewMetric()
+		mCalls.SetDataType(pdata.MetricDataTypeIntSum)
+		mCalls.SetName("calls")
+		mCalls.IntSum().DataPoints().Append(dpCalls)
+		mCalls.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mCalls)
+	}
+}
+
 // aggregateMetrics aggregates the raw metrics from the input trace data.
 // Each metric is identified by a key that is built from the service name
 // and span metadata such as operation, kind, status_code and any additional
 // dimensions the user has configured.
 func (p *processorImp) aggregateMetrics(traces pdata.Traces) {
-	// TODO: Add implementation
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		rspans := traces.ResourceSpans().At(i)
+		r := rspans.Resource()
+
+		attr, ok := r.Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+		serviceName := attr.StringVal()
+		p.aggregateMetricsForServiceSpans(rspans, serviceName)
+	}
+}
+
+func (p *processorImp) aggregateMetricsForServiceSpans(rspans pdata.ResourceSpans, serviceName string) {
+	ilsSlice := rspans.InstrumentationLibrarySpans()
+	for j := 0; j < ilsSlice.Len(); j++ {
+		ils := ilsSlice.At(j)
+		spans := ils.Spans()
+		for k := 0; k < spans.Len(); k++ {
+			span := spans.At(k)
+			p.aggregateMetricsForSpan(serviceName, span)
+		}
+	}
+}
+
+func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Span) {
+	latencyInMilliseconds := float64(span.EndTime()-span.StartTime()) / float64(time.Millisecond.Nanoseconds())
+
+	// Binary search to find the latencyInMilliseconds bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
+
+	p.lock.Lock()
+	key := buildKey(serviceName, span, p.dimensions)
```

Comment:
Nit: I think this line can be before the lock, I do not see it touching any shared data.